### PR TITLE
React router update

### PR DESCRIPTION
**Pull request by Matt-Eva**

### CODE REVIEW

**Review by lizbur10 — CHANGES_REQUESTED** [review]
This is looking good. A few things:

1. It doesn't really make sense to me that a list of MovieCard components is rendered on the Home page and then students are asked to add code to render the same info on the Movie page. I suggest just having students render a list of movie titles on the home page and then use the UserCard on the Movie page
2. It might be a good idea to mention that translating titles into routes (and back) will be handled automatically so students don't need to worry about it.
3. When I run the tests I'm getting an error about "/bad-route" not being found (in the test of the error page). The tests are all passing but it would probably be good if we can find a way suppress or clear that error.
4. The tests are also throwing warnings about using the regex global modifier. I just removed them from my test files and it worked fine so I'll do that here as well.

File: src/__tests__/Actors.test.js [hunk 1]
```diff
@@ -1,28 +1,33 @@
...
 
   expect(errorSpy).not.toHaveBeenCalled();
 
   errorSpy.mockRestore();
 });
 
 test("renders 'Actors Page' inside of the <h1 />", () => {
-  render(<Actors />);
+  render(<RouterProvider router={router}/>);
   const h1 = screen.queryByText(/Actors Page/g);
```
Comment (lizbur10) [inline src/__tests__/Actors.test.js:24]:
```suggestion
  const h1 = screen.queryByText(/Actors Page/);
```

File: src/__tests__/Directors.test.js [hunk 2]
```diff
@@ -1,28 +1,33 @@
...
 
   expect(errorSpy).not.toHaveBeenCalled();
 
   errorSpy.mockRestore();
 });
 
 test("renders 'Directors Page' inside of a <h1 />", () => {
-  render(<Directors />);
+  render(<RouterProvider router={router}/>);
   const h1 = screen.queryByText(/Directors Page/g);
```
Comment (lizbur10) [inline src/__tests__/Directors.test.js:24]:
```suggestion
  const h1 = screen.queryByText(/Directors Page/);
```

File: src/__tests__/Home.test.js [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
+import { RouterProvider, createMemoryRouter} from "react-router-dom";
+import routes from "../routes.js"
+import Home from "../pages/Home";
+
+const router = createMemoryRouter(routes)
 
 test("renders 'Home Page' inside of a <h1 />", () => {
-  render(<Home />);
+  render(<RouterProvider router={router}/>);
   const h1 = screen.queryByText(/Home Page/g);
```
Comment (lizbur10) [inline src/__tests__/Home.test.js:11]:
```suggestion
  const h1 = screen.queryByText(/Home Page/);
```

File: src/__tests__/Movie.test.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  render(<RouterProvider router={router} />);
+
+  expect(errorSpy).not.toHaveBeenCalled();
+
+  errorSpy.mockRestore();
+});
+
+test("renders movie's title in an h1", () => {
+  render(<RouterProvider router={router} />);
+  const h1 = screen.queryByText(/Doctor Strange/g);
```
Comment (lizbur10) [inline src/__tests__/Movie.test.js:26]:
```suggestion
  const h1 = screen.queryByText(/Doctor Strange/);
```

File: src/__tests__/Movie.test.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  expect(h1).toBeInTheDocument();
+  expect(h1.tagName).toBe("H1");
+});
+
+test("renders movie's time within a p tag", () => {
+  render(<RouterProvider router={router} />);
+  const p = screen.queryByText(/115/g);
```
Comment (lizbur10) [inline src/__tests__/Movie.test.js:33]:
```suggestion
  const p = screen.queryByText(/115/);
```

File: src/__tests__/NavBar.test.js [hunk 6]
```diff
@@ -30,18 +30,6 @@ test("renders a Home <NavLink>", async () => {
...
-  expect(a.tagName).toBe("A");
-  expect(a.href).toContain("/");
-
-  fireEvent.click(a, { button: 0 });
-
-  expect(a.classList).toContain("active");
-});
-
 test("renders a Actors <NavLink>", async () => {
   const a = screen.queryByText(/Actors/g);
```
Comment (lizbur10) [inline src/__tests__/NavBar.test.js:34]:
```suggestion
  const a = screen.queryByText(/Actors/);
```

File: src/__tests__/index.test.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  );
+  expect(document.querySelector(".navbar")).toBeInTheDocument();
+});
+
+test('renders the Home component on route "/"', () => {
+  const router = createMemoryRouter(routes)
+  render(
+    <RouterProvider router={router}/>
+);
+  expect(screen.getByText(/Home Page/g)).toBeInTheDocument();
```
Comment (lizbur10) [inline src/__tests__/index.test.js:19]:
```suggestion
  expect(screen.getByText(/Home Page/)).toBeInTheDocument();
```

File: src/__tests__/index.test.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+test('renders the Actors component on route "/actors"', () => {
+    const router = createMemoryRouter(routes, {
+        initialEntries: ['/actors'],
+        initialIndex: 0
+    })
+  render(
+    <RouterProvider router={router}/>
+);
+  expect(screen.getByText(/Actors Page/g)).toBeInTheDocument();
```
Comment (lizbur10) [inline src/__tests__/index.test.js:30]:
```suggestion
  expect(screen.getByText(/Actors Page/)).toBeInTheDocument();
```

File: src/__tests__/index.test.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+test('renders the Directors component on route "/directors"', () => {
+    const router = createMemoryRouter(routes, {
+        initialEntries: ['/directors'],
+        initialIndex: 0
+    })
+  render(
+      <RouterProvider router={router}/>
+  );
+  expect(screen.queryByText(/Directors Page/g)).toBeInTheDocument();
```
Comment (lizbur10) [inline src/__tests__/index.test.js:41]:
```suggestion
  expect(screen.queryByText(/Directors Page/)).toBeInTheDocument();
```

File: src/__tests__/index.test.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+test('renders the Movies component on route "/movie/:title"', () => {
+    const title = "Doctor Strange"
+    const router = createMemoryRouter(routes, {
+        initialEntries: [`/movie/${title}`],
+        initialIndex: 0
+    })
+  render(
+    <RouterProvider router={router}/>
+);
+  expect(screen.queryByText(/Doctor Strange/g)).toBeInTheDocument();
```
Comment (lizbur10) [inline src/__tests__/index.test.js:53]:
```suggestion
  expect(screen.queryByText(/Doctor Strange/)).toBeInTheDocument();
```

File: src/__tests__/index.test.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+test("renders an error page when given a bad URL", () =>{
+  const router = createMemoryRouter(routes, {
+      initialEntries: ["/bad-route"],
+      initialIndex: 0
+  })
+  render(
+      <RouterProvider router={router} />
+  )
+  expect(screen.getByText(/Oops! Looks like something went wrong./g)).toBeInTheDocument()
```
Comment (lizbur10) [inline src/__tests__/index.test.js:64]:
```suggestion
  expect(screen.getByText(/Oops! Looks like something went wrong./)).toBeInTheDocument()
```

**Review by lizbur10 — COMMENTED** [review]
Removing global flag from test files

**Comment by Matt-Eva** [comment]
Thanks for the feedback, Liz! I think those are all good changes.

1. I agree that we should probably just render the titles to our Home Page, along with a link directing a user to view info (I've learned that it's accessibility best practice to not nest links within header tags). I do think it would still make more sense to use MovieCards on Home to render the list of titles, as Card components seem to be traditionally used to render lists of information. Using it on the Movie Page component might be a little redundant, since that page is only rendering info about a single movie anyway. 
2. Good call! I'll add a note about URL encoding.
3. Thanks for flagging. I'll look into that and see if I can replicate the error.
4. I did notice that when I was changing the tests. Thank you for fixing that!

**Comment by Matt-Eva** [comment]
Ok, I've been coding through the solution, and I've found some problems.

1. Our tests aren't currently checking for MovieCards at all. We pass all of the tests if we just render `Home Page` on the home page route. So we'll need to add some tests for both of these components, I think.
2. Our tests aren't currently checking of either NavBar or our ErrorPage are rendering for every route - they just check if they're being rendered on our Home Route.
3. @lizbur10 R.e. the error you received regarding "/bad-route": Did it look like this?

> `  console.warn
      No routes matched location "/bad-route" `

That's what I see even with all the tests passing.

4. I'm also seeing this console message pop up for our Home test file: 

> `A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.`

I'm not sure what this is referring to, but can keep digging into it.

**Review by lizbur10 — CHANGES_REQUESTED** [review]
This is looking really great! A made a few more very minor suggestions (mostly punctuation 🙂 )

File: README.md [hunk 12]
```diff
@@ -2,96 +2,131 @@
...
 
 ```txt
 src/
 ├── data.js
 ├── index.js
-└── components/
-    ├── App.js
+├── routes.js
+├── components/
+    ├── NavBar.js
```
Comment (Matt-Eva) [inline README.md:28]:
```suggestion
    ├── MovieCard.js
```

File: README.md [hunk 13]
```diff
@@ -2,96 +2,131 @@
...
+    ├── MovieCard.js
```
Comment (Matt-Eva) [inline README.md:29]:
```suggestion
    ├── NavBar.js
```

File: README.md [hunk 14]
```diff
@@ -2,96 +2,131 @@
 
 ## Learning Goals
 
-- Use the `<Route>` component to display different components based on the URL
+- Use `createBrowserRouter` to create a client-side router
```
Comment (AlveeM) [inline README.md:5]: Add a `.` after each Learning Goal for accessibility.

Comment (Matt-Eva) [inline README.md:5]:
```suggestion
- Use `createBrowserRouter` to create a client-side router.
```

File: README.md [hunk 15]
```diff
@@ -2,96 +2,131 @@
...
+- Use `RouterProvider` to include the router in your app
```
Comment (Matt-Eva) [inline README.md:6]:
```suggestion
- Use `RouterProvider` to include the router in your app.
```

File: README.md [hunk 16]
```diff
@@ -2,96 +2,131 @@
...
 - Use the `<NavBar>` component to allow client-side navigation
```
Comment (Matt-Eva) [inline README.md:7]:
```suggestion
- Use the `<NavBar>` component to allow client-side navigation.
```

File: README.md [hunk 17]
```diff
@@ -2,96 +2,131 @@
...
+- Use `errorElement` to set up router error handling
```
Comment (Matt-Eva) [inline README.md:8]:
```suggestion
- Use `errorElement` to set up router error handling.
```

**Pull request merged**